onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -2550,19 +2550,10 @@ def aten_diagonal(self: TReal, offset: int = 0, dim1: int = 0, dim2: int = 1) ->
     # This is because computing diagonal sum is on dim2 after transpose by perm
     axes = [self_rank - 2]
 
-    return _aten_diagonal_onnx(self, offset, dim1, dim2, perm, axes)
-
-
-@torch_op("aten::diagonal", private=True, traceable=True)
-def _aten_diagonal_onnx(
-    self: TTensor, offset: int, dim1: int, dim2: int, perm: Sequence[int], axes: Sequence[int]
-) -> TTensor:
-    neg_1 = op.Constant(value_ints=[-1])
-    dim1_size = op.Reshape(op.Gather(op.Shape(self), dim1), neg_1)  # row
-    dim2_size = op.Reshape(op.Gather(op.Shape(self), dim2), neg_1)  # col
+    dim1_size = op.Shape(self, end=dim1, start=dim1 + 1)  # row
```

Comment:
Can `dim1` be negative? Especially -1? Same for `dim2`.

Reply:
ah it can. Thanks for pointing this out!

Reply:
Reverted to use gather. Added a comment

Reply:
Hmm but its handled in 
![image](https://github.com/user-attachments/assets/81f3225f-b6d3-483e-a2d4-97a3645351f5)